cmd/util/ledger/migrations/multiple_contract_migration.go

```diff
@@ -0,0 +1,350 @@
+	if len(key.KeyParts) != 3 ||
+		key.KeyParts[0].Type != state.KeyPartOwner ||
+		key.KeyParts[1].Type != state.KeyPartController ||
+		key.KeyParts[2].Type != state.KeyPartKey {
+		return flow.RegisterID{}, fmt.Errorf("key not in expected format %s", key.String())
+	}
+
+	return flow.NewRegisterID(string(key.KeyParts[0].Value), string(key.KeyParts[1].Value), string(key.KeyParts[2].Value)), nil
+}
+
+func createContractNamesKey(originalKey ledger.Key) ledger.Key {
+	return ledger.Key{
+		KeyParts: []ledger.KeyPart{
+			originalKey.KeyParts[0],
+			originalKey.KeyParts[1],
+			{
+				Type:  state.KeyPartKey,
+				Value: []byte("contract_names"),
+			},
+		},
+	}
+}
+
+func encodeContractNames(contractNames []string) ([]byte, error) {
+	sort.Strings(contractNames)
+	var buf bytes.Buffer
+	cborEncoder := cbor.NewEncoder(&buf)
+	err := cborEncoder.Encode(contractNames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot encode contract names")
+	}
+	return buf.Bytes(), nil
+}
+
+func contractsRegister(contractsKey ledger.Key, contractNames []string) (ledger.Payload, error) {
+	encodedContract, err := encodeContractNames(contractNames)
+	if err != nil {
+		return ledger.Payload{}, err
+	}
+	return ledger.Payload{
+		Key:   createContractNamesKey(contractsKey),
+		Value: encodedContract,
+	}, nil
+}
+
+func migrateRegister(p ledger.Payload) ([]ledger.Payload, error) {
+	registerId, err := keyToRegisterId(p.Key)
+	if err != nil {
+		return nil, err
+	}
+	if !isAddress(registerId) {
+		return []ledger.Payload{p}, nil
+	}
+
+	switch registerId.Key {
+	case "code":
+		if em, hasEM := MultipleContractsSpecialMigrations[registerId.Owner]; hasEM {
+			log.Info().
+				Err(err).
+				Str("address", flow.BytesToAddress([]byte(registerId.Owner)).HexWithPrefix()).
+				Msg("Using exceptional migration for address")
+			return em(p)
+		}
+		return migrateContractCode(p)
+	case "contract":
+		return migrateContract(p)
+	default:
+		return []ledger.Payload{p}, nil
+	}
+}
+
+func migrateContract(p ledger.Payload) ([]ledger.Payload, error) {
+	address := common.BytesToAddress(flow.BytesToAddress(p.Key.KeyParts[0].Value).Bytes())
+	storedData, version := interpreter.StripMagic(p.Value)
+	if len(storedData) == 0 {
+		return []ledger.Payload{}, nil
+	}
+	storedValue, err := interpreter.DecodeValue(storedData, &address, []string{"contract"}, version)
+	if err != nil {
+		log.Error().
+			Err(err).
+			Str("address", address.Hex()).
+			Msg("Cannot decode contract at address")
+		return nil, err
+	}
+
+	value := interpreter.NewSomeValueOwningNonCopying(storedValue).Value.(*interpreter.CompositeValue)
+	pieces := strings.Split(string(value.TypeID), ".")
+	if len(pieces) != 3 {
+		log.Error().
+			Str("TypeId", string(value.TypeID)).
+			Str("address", address.Hex()).
+			Msg("contract TypeId not in correct format")
+		return nil, fmt.Errorf("contract TypeId not in correct format")
+	}
+	newKey := changeKey(p.Key, fmt.Sprintf("contract\x1F%s", pieces[2]))
```

Comment:
Maybe `formatContractKey` from Cadence's `runtime` package should be used?

Reply:
but thats private? should we export it for a single use migration?

Reply:
Right, didn't see that, and not worth it to export for this one-off migration